lib/resty/upstream/healthcheck.lua

```diff
@@ -134,7 +138,15 @@ local function peer_fail(ctx, is_backup, id, peer)
     -- print("ctx fall: ", ctx.fall, ", peer down: ", peer.down,
           -- ", fails: ", fails)
 
-    if not peer.down and fails >= ctx.fall then
+    local u_key = gen_upstream_key(u,"all_down")
```

Comment:
Style: needs a space after the comma.

Also, it is a bit sad to introduce a new string concat and creation operation here. Given that upstream should be fixed, can we recycle this key string instead of trying to generate a new one every time?

Reply:
And it's a bit confusing to call the "all_down" key an "upstream key".

Reply:
Fixed style issues. 

gen_upstream_key() is used 5 more times exactly like this. It does simple concatenation. So, code looks cleaner with it. I made it with the same idea as already existing gen_peer_key() . Just use suffix instead of prefix. 

Not sure how to call this key then. Maybe, all_peers_down ?

Reply:
I removed gen_upstream_key() from peer_fail(). Instead added extra arg all_peers_down_key.
Since peer_fail() is called only once from peer_error() function and we already have this key in peer_error(). 
Please, check if you find this better? 

Now peer_fail() and peer_ok() have different number of arguments, which doesn't look very nice I think. 

Since this key is global for an upstream it is possible to call gen_upstream_key() once in do_check(). But, then we need to extend all the following functions (check_peers(), check_peer() .. ) with this extra arg. Not sure, what is better.